### PR TITLE
server,cli/start: gracefully drain tenant servers upon service disable

### LINKED ISSUE #107856: ccl/serverccl: TestServiceShutdownUsesGracefulDrain failed
ccl/serverccl.TestServiceShutdownUsesGracefulDrain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11115347?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11115347?buildTab=artifacts#/) on master @ [f295bd861a3a427652b19c2254d2401ebb4a3c8e](https://github.com/cockroachdb/cockroach/commits/f295bd861a3a427652b19c2254d2401ebb4a3c8e):

```
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !goroutine 27787625 [sync.Cond.Wait]:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.runtime_notifyListWait(0xc013189690, 0x20e)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/runtime/sema.go:517 +0x14c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.(*Cond).Wait(0x58c8f60?)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/sync/cond.go:70 +0x8c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0029f58f0, {0x79c1540, 0xc021bb24b0}, {0x79a4c20, 0xc003768a80}, 0xc003870a80)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:378 +0x318
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x79c1540?, 0xc021bb24b0?})
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !goroutine 27787367 [sync.Cond.Wait]:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.runtime_notifyListWait(0xc013189610, 0x356)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/runtime/sema.go:517 +0x14c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !sync.(*Cond).Wait(0x58c8f60?)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	GOROOT/src/sync/cond.go:70 +0x8c
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0029f56c0, {0x79c1540, 0xc02de2fad0}, {0x79a4c20, 0xc003768a80}, 0xc003870a80)
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:378 +0x318
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x79c1540?, 0xc02de2fad0?})
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !For more context, check log files in: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServiceShutdownUsesGracefulDrain2336992875
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !****************************************************************************
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !This node experienced a fatal error (printed above), and as a result the
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !process is terminating.
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !Fatal errors can occur due to faulty hardware (disks, memory, clocks) or a
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !problem in CockroachDB. With your help, the support team at Cockroach Labs
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !will try to determine the root cause, recommend next steps, and we can
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !improve CockroachDB based on your report.
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !Please submit a crash report by following the instructions here:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !    https://github.com/cockroachdb/cockroach/issues/new/choose
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !If you would rather not post publicly, please contact us directly at:
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !    support@cockroachlabs.com
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !
F230730 06:57:00.234543 28283619 server/server_sql.go:1682  [T2,n1] 1 !The Cockroach Labs team appreciates your feedback.
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multi-tenant @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServiceShutdownUsesGracefulDrain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-30222
Epic: CRDB-26691

### PR DESCRIPTION
**Pull request by knz**
First commit from #108612.
Thanks to @stevendanna for brainstorming solutions.

TLDR: when the service mode of a secondary tenant is changed to NONE,
but the data state is still READY, the tenant servers will now shut
down gracefully. This applies both to tenant servers run as separate
processes (via `cockroach mt start-sql`) and those started via the
internal server controller.

More details follow.

For context, the reader is invited to first refer to the commit
immediately prior.

Prior to this patch, all internal shutdown requests were considered to
be non-graceful. This included the request that was triggered as a
result of a change in the service mode.

This caused two separate bugs:

- it created a race condition between the server controller (for
  shared-process servers) and the service mode checker task inside the
  SQL server (`startCheckService`). When the service mode was flipped to
  'NONE', the server controller initiated a graceful drain but then that
  was caught up by the service mode checker and converted to a
  hard shutdown. This particular bug was a regression caused by #96144
  and had been identified through flakes in `TestServiceShutdownUsesGracefulDrain`.

- it was generally causing shutdowns upon changes in service mode to
  always be run as hard shutdowns (both for external-process and shared-process
  ervers). This is generally undesirable when the data state is still
  READY, as it prevents cleaning up the SQL liveness record and can
  cause delays during subsequent server startups.

This patch fixes it by using a graceful shutdown request whenever
the service mode changes and the data state is still READY.
Since the tenant server now handles graceful drains, the patch
also removes this responsibility from the server controller.

Release note: None
Fixes #107856
Epic: CRDB-26691

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
I like where this landed and it ended up being simpler than the approach I took.  

Do we think we need any sort of timeout on the graceful drain here such that after some amount of time we give up draining gracefully and do a hard shutdown?

**Comment by knz** [comment]
> Do we think we need any sort of timeout on the graceful drain here such that after some amount of time we give up draining gracefully and do a hard shutdown?

I am ambivalent about it. It could go either way, but I'd like to emphasize that this is orthogonal to the issue that this PR is fixed. If we want to add such a timeout, I'd rather do that in a different PR.

**Review by yuzefovich — COMMENTED** [review]
LGTM, but I'll defer to Steven.

> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108613#-:-Nbeoy9_7CINJWCCeCM_:b-v879n2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @knz, @rachitgsrivastava, and @stevendanna)

___
*[`pkg/server/server_sql.go` line 1768 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108613#-Nbeojdj3X6kdELo6rfG:-Nbeojdj3X6kdELo6rfH:bvv78zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6585818b01f2828747d062e0e9170f3d2b390cbe/pkg/server/server_sql.go#L1768)):*
> ```Go
> 		return false, errors.AssertionFailedf("service check failed: expected data state %v, record says %v", expected, actual)
> 	}
> 	return true, nil
> ```

nit: I was a bit confused by this, consider leaving a comment here that given `err` return argument is `nil`, `useGracefulDrain` return argument doesn't matter.

**Review by stevendanna — APPROVED** [review]
>I am ambivalent about it. It could go either way, but I'd like to emphasize that this is orthogonal to the issue that this PR is fixing. If we want to add such a timeout, I'd rather do that in a different PR.

That's fine by me. I'm not sure it is really orthogonal since this PR effectively changes what was previously an immediate shutdown to a graceful shutdown and thus introduces the possibility of the shutdown taking a non-trivial amount of time.  But, since this is new functionality that hasn't been exposed to users, there is no existing user expectation.

**Comment by knz** [comment]
> I'm not sure it is really orthogonal since this PR effectively changes what was previously an immediate shutdown to a graceful shutdown

I believe that is incorrect. In the v23.1 code (the original version I implemented) the shutdown was graceful.
It only became immediate when I merged #96144, which we are considering a regression (only in the master branch for now). This patch recovers the v23.1 behavior.

**Comment by knz** [comment]
TFYR!

bors r=yuzefovich,stevendanna

**Comment by knz** [comment]
bors r-

i'll address yahor's nit first

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108613)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @rachitgsrivastava)

___
*[`pkg/server/server_sql.go` line 1768 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108613#-Nbeojdj3X6kdELo6rfG:-Nboz5Bj-epX0M7HIgAh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6585818b01f2828747d062e0e9170f3d2b390cbe/pkg/server/server_sql.go#L1768)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I was a bit confused by this, consider leaving a comment here that given `err` return argument is `nil`, `useGracefulDrain` return argument doesn't matter.
</blockquote></details>

Done.

**Comment by knz** [comment]
bors r=yuzefovich,stevendanna
